### ISSUE TITLE
Flag Markdown syntax inside HTML tags

### ISSUE THREAD

**Issue by guillp** [rvben/rumdl#801]
### Problem or motivation

Now that MD033 properly allows using arbitrary HTML tags inside container tags such as `<details>`, I'd like `rumdl` to report when Markdown syntax such as links, images, etc. are present inside those tags. This should be linted since Markdown syntax is actually unsupported inside HTML container tags, and will be rendered as simple text.

### Proposed solution

Either a new feature in MD033 or a new rule entirely, that parses the content between opening and closing HTML tags, and checks for presence of Markdown syntax.
If Markdown syntax is present (e.g `[Links](https://github.com)`, `**bold**`, etc.), report it as a warning.

### Alternatives considered

MD033 is quite related and I tried to implement this new feature on top of it. But given the current limitations described below in Additional Context, a new rule might be better.

But @rvben you know your codebase much better than me so I'm sure you'll find the best way to implement it :)

### Feature category

Improve existing rule

### Example markdown

```markdown
<details><summary>[FAIL](https://test.local)</summary>This Markdown link is rendered as plain text</details>
<details><summary><a href="https://test.local">OK</a></summary>This HTML link is rendered as expected</details>
```
Which renders as:

<details><summary>[FAIL](https://test.local)</summary>This Markdown link is rendered as plain text</details>
<details><summary><a href="https://test.local">OK</a></summary>This HTML link is rendered as expected</details>

### Contribution

- [ ] I would be willing to submit a PR to implement this feature

### Additional context

I investigated a bit:
- today `rumdl` parser does not consider the content between opening and closing HTML tags to be Markdown, so it does not parse any Markdown syntax like links or images that may be present in there, thus does not expose such elements for easy access in `LintContext`.
- so any rule that detects markdown syntax in there must parse the inside content of HTML tags itself (e.g. by creating a new `LintContext` on the text between the HTML tags.
- But I didn't find an already available way in `rumdl` to obtain that text content between matching HTML tags. Today tags themselves are available in `LintContext.html_tags()`, but with no relationship between them, thus some logic is needed to obtain the inside text

**Comment by rvben** [comment]
Sorry for the long silence. I ended up building the rule to find out whether it
was worth having, and the answer turned out to be no. Reasoning below, since you
put real work into the report.

One correction that's useful regardless: the dividing line is the HTML *block*,
not the tag pair. A blank line ends a block no matter which tags are still open,
so inside a single `<details>` a link can be dead while emphasis four lines later
renders fine. Which also means you don't need to extract the text between tags at
all. `LintContext::is_in_html_block(line)` already answers it, and it matches the
renderer line for line.

So I wrote MD089 and ran it over 33,000 doc-site files and 7,400 READMEs. One
true positive in total. READMEs are where I expected it to concentrate, since
that's where the `<div align="center">` banners live, and there it found nothing:
of the 16,388 lines sitting inside HTML blocks, 3,315 hold an `<a>` tag and
exactly one holds a markdown link. Someone who opens a centered div keeps writing
HTML. I checked the `<details>` plus code fence shape separately too, and got 68
occurrences with all 68 written correctly.

The reason is that this bug is loud. A markdown link inside an HTML block shows
up as visible `[text](url)` on the rendered page, usually in the most-viewed file
in the repo, so it gets caught on the first preview. That's the opposite of what
a linter is for. rumdl earns its keep on the things you can't see in the output.

Your report did turn up three real bugs though, and all three shipped in v0.2.54:
reference links inside an HTML block leaking into the link index (so MD059 fired
on one and not the other in the same position), a `<pre>` nested in a `<table>`
with a Prettier-split `</pre\n>` never closing and silencing every rule after it,
and `markdown="1"` handling being wrong in three separate ways, including reading
kramdown's `markdown="0"` as an opt-in rather than its opposite. Better return
than the rule itself would have been.

If you have a corpus where this shows up at any real rate, send me the numbers
and I'll reconsider. The rule is finished and sitting on a branch, so it's a
merge rather than a rebuild. I just don't want to ship it on the assumption that
it's common, having made that assumption myself and been wrong.

**Comment by guillp** [comment]
Hey @rvben
No apologies necessary! Thanks for your time and efforts investigating this. I was on holidays myself :)

About your analysis and metrics: I am not surprised by your findings since, as you say, this is a very visible issue when it happens, so they typically are fixed quickly. Plus using HTML formatting inside Markdown is relatively uncommon in the first place.

However, the rationale behind such checks is for future changes rather than for existing docs:
I recently reviewed a couple of PRs at work containing this issue. PRs in GitHub are tricky since they don't include Markdown rendering, so they don't make it easy to see the issue. And since a linter such as rumdl could automatically flag such issues, it could systematically reject the PR as part of CI, or even inform the writer that such issue is present,  with a readable explanation on how to fix it, before he commits or pushes.

Could you please publish your branch? I'll run it again our documentation base to see if it contains some more issues that may have slipped in.

**Comment by guillp** [comment]
@rvben : thanks for merging and releasing this in 4333acf54a400cb8283bb42e9513b464d57dbbf2 :)

You wanted numbers, here are some:

```
$ rumdl check --enable MD091
Rule Violation Statistics:
Rule     Violations   Fixable  Percentage
--------------------------------------------------
MD091    754          -         100.0%
--------------------------------------------------
Total    754          0         100.0%
```

Those are all in automatically generated files that contain tables of references.
To give you an overview of this docbase:

```
$ tokei -t Markdown,MDX
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 Language              Files        Lines         Code     Comments       Blanks
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 MDX                     105        18209            0        12750         5459
─────────────────────────────────────────────────────────────────────────────────
 Markdown                 81        11536            0         8258         3278
 |- BASH                   6           71           42           24            5
 |- JavaScript             1            1            1            0            0
 |- JSON                   1           26           26            0            0
 |- Markdown               2           42            0           32           10
 |- MDX                    2            6            0            4            2
 |- Shell                  2            8            8            0            0
 |- YAML                   1           12           11            1            0
 (Total)                            11702           88         8319         3295
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 Total                   186        29911           88        21069         8754
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

To be noted: MDX renders Markdown in HTML "correctly" (as in with the expected HTML tags).

**Comment by rvben** [comment]
You're right about MDX: Markdown inside JSX elements, including `<table>` and `<td>`, renders as Markdown. Those links should not trigger MD091.

This is fixed in [v0.2.68](https://github.com/rvben/rumdl/releases/tag/v0.2.68). The shared MDX parser now recognizes links in generated tables, so normal link rules check them too. JavaScript expressions, JSX attributes, comments, and code are excluded. I verified this against a public TypeDoc-generated reference table and the MDX renderer, including a compact version without blank lines inside cells.

Could you rerun `rumdl check --enable MD091` with v0.2.68? `.mdx` files select MDX automatically; `.md` files compiled as MDX need `flavor = "mdx"` or a per-file mapping. The earlier 754 warnings may include false positives, so the new count will be more useful.

**Comment by guillp** [comment]
With 0.2.68:
```
rumdl check --enable MD091

Success: No issues found in 192 files (49ms)
```

Thanks and good job!

**Comment by rvben** [comment]
Thanks for the verification, @guillp ! Let me know if you hit anything else.

**Comment by guillp** [comment]
I forgot to mention that MD039 now triggers on Markdown links embedded inside HTML, which is good! (15 occurrences in my doc base FYI)
